

<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Manual &#8212; Easy Protein Tree Annotating  documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/bizstyle.css" />
    
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Commandline Reference" href="commandline.html" />
    <link rel="prev" title="Quick Start" href="quickstart.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <!--[if lt IE 9]>
    <script src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="commandline.html" title="Commandline Reference"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="quickstart.html" title="Quick Start"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Easy Protein Tree Annotating  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Manual</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="manual">
<h1>Manual<a class="headerlink" href="#manual" title="Permalink to this headline">¶</a></h1>
<section id="before-you-start">
<h2>Before you start<a class="headerlink" href="#before-you-start" title="Permalink to this headline">¶</a></h2>
<section id="configuration-file">
<h3>Configuration file<a class="headerlink" href="#configuration-file" title="Permalink to this headline">¶</a></h3>
<p>A configuration file named ‘Config.ini’ can be found under the root path of the EPTA folder,
which stores all default command-related parameters. Users can edit it manually to change
the default settings of the tool.</p>
<p>Additionally, EPTA generates a ‘Config.ini’ file in the input file path after a run. Thus, if
one needs to run EPTA multiple times using the same input file, edit configuration file in the
input file path is also a option.</p>
</section>
<section id="log-file">
<h3>Log file<a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>EPTA automatically generates a log file that records everything that happened during a run.
Users can find the log under the output path of each run.</p>
<p>Also, a configuration log will be automatically generated in the output path that allows user
check or confirm configuration of a run.</p>
</section>
</section>
<section id="make-a-tree-from-fasta-file">
<h2>Make a tree from FASTA file<a class="headerlink" href="#make-a-tree-from-fasta-file" title="Permalink to this headline">¶</a></h2>
<p>An annotated phylogenetic tree can be built directly from FASTA file, and able to add protein ID, description,
organism lineage, protein domain are able to be added to the tree.</p>
<section id="example-command-line">
<h3>Example command-line<a class="headerlink" href="#example-command-line" title="Permalink to this headline">¶</a></h3>
<p>Under construction</p>
</section>
<section id="input-file">
<h3>Input file<a class="headerlink" href="#input-file" title="Permalink to this headline">¶</a></h3>
<p>A ‘-i’ flag is what we used to declare input file, which is obligatorily required. The inputted file can be a FASTA file
in plain text format, or contained in a gunzip or tar.gz compressed file, or a folder contains any the
aforementioned file types.</p>
<blockquote>
<div><p>Input a FASTA file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i test_fasta.fasta -o ./test
</pre></div>
</div>
<p>Input a gunzip file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i test_fasta.gz -o ./test
</pre></div>
</div>
<p>Input a tar.gz file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i test_fasta.tar.gz -o ./test
</pre></div>
</div>
<p>Input a folder:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-ls ./fasta_files
<span class="s1">&#39;test_fasta.fasta&#39;</span> <span class="s1">&#39;test_fasta.gz&#39;</span> <span class="s1">&#39;test_fasta.tar.gz&#39;</span>

epta -i ./fasta_files -o ./test
</pre></div>
</div>
</div></blockquote>
</section>
<section id="output-path">
<h3>Output path<a class="headerlink" href="#output-path" title="Permalink to this headline">¶</a></h3>
<p>A ‘-i’ flag is what we used to store output file, which is obligatory required.</p>
<p>Our tool will automatically detect whether the output path exists and if it is,
you will see the following caution:</p>
<blockquote>
<div><div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Output directory is existed, do you want to continue? (Y/N)
</pre></div>
</div>
<div class="admonition caution">
<p class="admonition-title">Caution</p>
<p>Once the user input <strong>Y</strong> and <strong>press enter</strong>, all tool-related files, including the previous dataframe,
parsed fasta file, etc., will be <strong>removed</strong> from that path.</p>
</div>
</div></blockquote>
<p>If you would like to skip manual confirmation, add <strong>-redo</strong> flag in your command line.</p>
</section>
<section id="check-point">
<h3>Check point<a class="headerlink" href="#check-point" title="Permalink to this headline">¶</a></h3>
<p>EPTA will automatically generates a check point after it starting a run. You will
see the following caution if any check point file has been found in the output path:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Output directory is existed, do you want to read check point? (Y/N)
</pre></div>
</div>
<p>To continue the run from the place where last run ended, input <strong>Y</strong> and press enter.</p>
<p>Same as output path check, if you would like to skip manual confirmation, add <strong>-redo</strong>
flag in your command line.</p>
</section>
<section id="parse-fasta">
<h3>Parse FASTA<a class="headerlink" href="#parse-fasta" title="Permalink to this headline">¶</a></h3>
<section id="sequence-recognition">
<h4>Sequence recognition<a class="headerlink" href="#sequence-recognition" title="Permalink to this headline">¶</a></h4>
<p>We introduced Biopython module to distinguish sequence info. By default, protein name and ID
will be extracted from sequence headers automatically. Users do not need any command-line flags to make
it happen.</p>
</section>
<section id="dataframe">
<h4>Dataframe<a class="headerlink" href="#dataframe" title="Permalink to this headline">¶</a></h4>
<p>Our tool will generate a human-readable dataframe called ‘info_index.tsv’ in tsv format under the output
path after parse all inputted files. Furthermore, if users choose to run Pfamscan, an independent tsv
file named ‘Pfamscan_result.tsv’ will be generated under the output path in order to inspect all
domain-related data.</p>
<p>For a line in the dataframe file, index is a unique random ID assigned by EPTA, and columns are arranged
as:</p>
<blockquote>
<div><table class="docutils align-default">
<colgroup>
<col style="width: 7%" />
<col style="width: 5%" />
<col style="width: 5%" />
<col style="width: 8%" />
<col style="width: 13%" />
<col style="width: 23%" />
<col style="width: 23%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Random ID</p></th>
<th class="head"><p>Header</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>Name</p></th>
<th class="head"><p>Organism Lineage</p></th>
<th class="head"><p>Domain Information[database]</p></th>
<th class="head"><p>Domain Information[Pfamscan]</p></th>
<th class="head"><p>Pfamscan JSON data</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>If a parameter is not requested, the following columns will move forward in order.</p>
</section>
<section id="duplicate-headers">
<h4>Duplicate headers?<a class="headerlink" href="#duplicate-headers" title="Permalink to this headline">¶</a></h4>
<p>If you want to keep all duplicate sequence in your file, add <strong>-dh</strong> command to your
command line.</p>
</section>
<section id="local-data">
<h4>Local data?<a class="headerlink" href="#local-data" title="Permalink to this headline">¶</a></h4>
<p>Local sequences that contained in FASTA file need to be marked with a <strong>‘lcl|’</strong> identifier, in order to extract
information from it <strong>automatically</strong>. Protein name, ID, and taxonomy can be directly included in a header with specific
identifiers and splited with a vertical bar (‘|’), while the order of their <strong>arrangements doesn’t matter</strong>. And,
Pfamscan data should be included in another file if you do not want to fetch domain information or run Pfamscan
by EPTA, and the format of domain information should be in a JSON format which is the same with Pfamscan’s result.</p>
<blockquote>
<div><p>Parameter identifiers:</p>
<blockquote>
<div><table class="docutils align-default">
<colgroup>
<col style="width: 45%" />
<col style="width: 55%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Identifiers</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Name</p></td>
<td><p>[NAME]</p></td>
</tr>
<tr class="row-odd"><td><p>ID</p></td>
<td><p>[ID]</p></td>
</tr>
<tr class="row-even"><td><p>Taxonomy</p></td>
<td><p>[TAXON]</p></td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Example of a local sequence:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span><span class="linenos">1</span>  &gt;lcl| [ID]]OAO11745.1 | [NMAE]hydrogenase | [TAXON] cellular organisms; Eukaryota; Sar; Stramenopiles; Bigyra; Opalozoa; Opalinata; Blastocystidae; Blastocystis; Blastocystis sp. subtypes; Blastocystis sp. subtype 1
<span class="linenos">2</span>  MLSRLSRIATTKSMLVMNAARSFAAEAQGKLVSVKINGNEYKVPEGMTVLEACQAQGIHVPFVCHHPRLKPLGKCRVCVVEIRGDEFPIKTSCNTKVEEGMDIWTNSPKARSASNEALKTLMAGTPIDTKFKTMEMDEVLTESADDCYALHRDMSRCVDCKRCARACSELQGMNVLENNPQEGGFPVVPTGYHLLKDTECISCGQCNVVCPTGAIVEQSHIPRVKQAMKAGKVMVMQTAPATRVAFGENFGREPGEITTGKMIACAKALGFQYVFDTNFGADMTIMEEGTELLERIKNNGPFPMFTSCCPGWVNMAEKCYPEILPNLSSCRSPHMMVGSTLKTYWAKKMNLKPEDIYVVSLMPCTAKKDEIERKNMWLDEKTPFVDAVLTTKELGDFCKQEGITNWDNMAEMPFDTPLGTSSGAGDIFGVSGGVMEAALRTAYQLQTGKPLEKIVVDEARGLDGTKRFSVDMNGKKINCAVVHSKFARELVENVKAGKEDLQFVEVMACPGGCISGGGQPHSNRADTIEKRMNAIYKIDAGKTLRRSMDNPEIQTLYKEFFEKPNSHKAHELLHTTYAPQYVRSREEKVEEPAEAGGEGGEGSDVGEDGVTILYGSETGTTAKAAKALQSKFKAAGISSAVIPMNKIDVESLPEKKKLVLMTCTYGAGEFPAMAQEFWENLSDESLDDDFLEGVEFGVFGLGSKAFKQFCEAAHQLDERMEELGAERVVDCGEGNEKDPQQYKTAFEPWSKEAVEAFK
</pre></div>
</div>
<p>Example of a domain information format:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span><span class="linenos">1</span>  &gt;lcl| [ID]]OAO11745.1
<span class="linenos">2</span>  [{&quot;model_length&quot;:&quot;82&quot;,&quot;align&quot;:[&quot;#HMM       pvtltfDGkevtvpeGdtvasAllangvdvprsckygrprgelcaggeCrnClVeveg..epnvracstpvedGlkvetqt&quot;,&quot;#MATCH      v+++++G+e++vpeG tv++A++a+g++vp  c++      l++ g+Cr C+Ve+ g   p   +c t+ve+G+ + t++&quot;,&quot;#PP        6899********************************....789*************86215556*************9986&quot;,&quot;#SEQ       LVSVKINGNEYKVPEGMTVLEACQAQGIHVPFVCHH----PRLKPLGKCRVCVVEIRGdeFPIKTSCNTKVEEGMDIWTNS&quot;],&quot;env&quot;:{&quot;to&quot;:&quot;108&quot;,&quot;from&quot;:&quot;29&quot;},&quot;name&quot;:&quot;Fer2_4&quot;,&quot;acc&quot;:&quot;PF13510.6&quot;,&quot;sig&quot;:1,&quot;evalue&quot;:&quot;1.4e-17&quot;,&quot;desc&quot;:&quot;2Fe-2S iron-sulfur cluster binding domain&quot;,&quot;hmm&quot;:{&quot;to&quot;:&quot;81&quot;,&quot;from&quot;:&quot;3&quot;},&quot;act_site&quot;:null,&quot;type&quot;:&quot;Domain&quot;,&quot;bits&quot;:&quot;63.5&quot;,&quot;clan&quot;:&quot;CL0486&quot;,&quot;seq&quot;:{&quot;to&quot;:&quot;107&quot;,&quot;from&quot;:&quot;31&quot;,&quot;name&quot;:&quot;EMBOSS_001&quot;}},{&quot;model_length&quot;:&quot;52&quot;,&quot;align&quot;:[&quot;#HMM       rCigCgaCvaaCp....vkaieldeeenekgt.....ekveidpekClgCgaCvavCPtga&quot;,&quot;#MATCH     rC+ C++C++aC+    ++++e +++  e g+      +  ++ ++C++Cg+C  vCPtga&quot;,&quot;#PP        8***********87544555555543..333433322333667899*************98&quot;,&quot;#SEQ       RCVDCKRCARACSelqgMNVLENNPQ--EGGFpvvptGYHLLKDTECISCGQCNVVCPTGA&quot;],&quot;env&quot;:{&quot;to&quot;:&quot;214&quot;,&quot;from&quot;:&quot;156&quot;},&quot;name&quot;:&quot;Fer4_7&quot;,&quot;acc&quot;:&quot;PF12838.7&quot;,&quot;sig&quot;:1,&quot;evalue&quot;:&quot;3.2e-06&quot;,&quot;desc&quot;:&quot;4Fe-4S dicluster domain&quot;,&quot;hmm&quot;:{&quot;to&quot;:&quot;52&quot;,&quot;from&quot;:&quot;1&quot;},&quot;act_site&quot;:null,&quot;type&quot;:&quot;Domain&quot;,&quot;bits&quot;:&quot;27.6&quot;,&quot;clan&quot;:&quot;CL0344&quot;,&quot;seq&quot;:{&quot;to&quot;:&quot;214&quot;,&quot;from&quot;:&quot;156&quot;,&quot;name&quot;:&quot;EMBOSS_001&quot;}},{&quot;model_length&quot;:&quot;243&quot;,&quot;align&quot;:[&quot;#HMM       kvvvqvAPavrvalgeefglsv.aattgklvaalrklGfdkVfdtafgadltimeeasellerleeeeklpmitScCPgwvkyvekkypellpnlssvkSPqqilgaliKkylaek.....ekivvVsimPCtaKklEaareefksag..rdvDavlTtrElaellkeagidl.akleeeeldnplgessgagki...................................egvkeaevelegktlkvavvnGlknikklleklkageakydfiEvmaCpgGCigGg&quot;,&quot;#MATCH      +v+q+APa+rva+ge+fg ++ + ttgk++a+ + lGf++Vfdt+fgad+timee++eller++++ + pm+tScCPgwv+++ek ype+lpnlss++SP++++g+ +K+y+a+k     e+i+vVs+mPCtaKk E++r+++  ++    vDavlTt+El++  k++gi+  +++ e ++d+plg+ssgag i                                   +g+k+ +v+++gk++++avv+  k +++l+e++kag+ + +f+EvmaCpgGCi Gg&quot;,&quot;#PP        689***************************************************************99****************************************************************************9855579*******************9655****************************************************************************96.889*****************************8&quot;,&quot;#SEQ       VMVMQTAPATRVAFGENFGREPgEITTGKMIACAKALGFQYVFDTNFGADMTIMEEGTELLERIKNNGPFPMFTSCCPGWVNMAEKCYPEILPNLSSCRSPHMMVGSTLKTYWAKKmnlkpEDIYVVSLMPCTAKKDEIERKNMWLDEktPFVDAVLTTKELGDFCKQEGITNwDNMAEMPFDTPLGTSSGAGDIfgvsggvmeaalrtayqlqtgkplekivvdearglDGTKRFSVDMNGKKINCAVVHS-KFARELVENVKAGKEDLQFVEVMACPGGCISGG&quot;],&quot;env&quot;:{&quot;to&quot;:&quot;517&quot;,&quot;from&quot;:&quot;232&quot;},&quot;name&quot;:&quot;Fe_hyd_lg_C&quot;,&quot;acc&quot;:&quot;PF02906.14&quot;,&quot;sig&quot;:1,&quot;evalue&quot;:&quot;7.5e-87&quot;,&quot;desc&quot;:&quot;Iron only hydrogenase large subunit, C-terminal domain&quot;,&quot;hmm&quot;:{&quot;to&quot;:&quot;243&quot;,&quot;from&quot;:&quot;2&quot;},&quot;act_site&quot;:null,&quot;type&quot;:&quot;Domain&quot;,&quot;bits&quot;:&quot;291.0&quot;,&quot;clan&quot;:&quot;No_clan&quot;,&quot;seq&quot;:{&quot;to&quot;:&quot;517&quot;,&quot;from&quot;:&quot;233&quot;,&quot;name&quot;:&quot;EMBOSS_001&quot;}},{&quot;model_length&quot;:&quot;56&quot;,&quot;align&quot;:[&quot;#HMM       dvrkkRakalykiDkkkklrkSheNpevkklYkeflgeplsekahelLHThYtd&quot;,&quot;#MATCH     d+++kR++a+ykiD+ k+lr+S +Npe++ lYkef+++p+s+kahelLHT+Y +&quot;,&quot;#PP        789*************************************************76&quot;,&quot;#SEQ       DTIEKRMNAIYKIDAGKTLRRSMDNPEIQTLYKEFFEKPNSHKAHELLHTTYAP&quot;],&quot;env&quot;:{&quot;to&quot;:&quot;580&quot;,&quot;from&quot;:&quot;525&quot;},&quot;name&quot;:&quot;Fe_hyd_SSU&quot;,&quot;acc&quot;:&quot;PF02256.17&quot;,&quot;sig&quot;:1,&quot;evalue&quot;:&quot;6.2e-24&quot;,&quot;desc&quot;:&quot;Iron hydrogenase small subunit&quot;,&quot;hmm&quot;:{&quot;to&quot;:&quot;55&quot;,&quot;from&quot;:&quot;2&quot;},&quot;act_site&quot;:null,&quot;type&quot;:&quot;Domain&quot;,&quot;bits&quot;:&quot;83.8&quot;,&quot;clan&quot;:&quot;No_clan&quot;,&quot;seq&quot;:{&quot;to&quot;:&quot;579&quot;,&quot;from&quot;:&quot;526&quot;,&quot;name&quot;:&quot;EMBOSS_001&quot;}},{&quot;model_length&quot;:&quot;143&quot;,&quot;align&quot;:[&quot;#HMM       ilYgSetGnteklAkqlaeelgehgfnadvvslsdydeslseieeealllvvtsTfgnGdppengesffqdllelkgdeledgdlsgvrfavfglGdsayenFcaagkkldekleelGaerllkllegdednqegqeeafrkW&quot;,&quot;#MATCH     ilYgSetG+t k Ak+l+ +++++g+++ v+ ++++d  +++++e++ l++ t T+g G++p  ++ +f++ l +++  l+d+ l+gv+f vfglG++a ++Fc+a+++lde++eelGaer++  +eg+e++ + +++af+ W&quot;,&quot;#PP        89***********************************..68*********************7777.8888888885..67777***********************************************99********99&quot;,&quot;#SEQ       ILYGSETGTTAKAAKALQSKFKAAGISSAVIPMNKID--VESLPEKKKLVLMTCTYGAGEFPAMAQ-EFWENLSDES--LDDDFLEGVEFGVFGLGSKAFKQFCEAAHQLDERMEELGAERVVDCGEGNEKDPQQYKTAFEPW&quot;],&quot;env&quot;:{&quot;to&quot;:&quot;749&quot;,&quot;from&quot;:&quot;612&quot;},&quot;name&quot;:&quot;Flavodoxin_1&quot;,&quot;acc&quot;:&quot;PF00258.25&quot;,&quot;sig&quot;:1,&quot;evalue&quot;:&quot;1.5e-33&quot;,&quot;desc&quot;:&quot;Flavodoxin&quot;,&quot;hmm&quot;:{&quot;to&quot;:&quot;143&quot;,&quot;from&quot;:&quot;1&quot;},&quot;act_site&quot;:null,&quot;type&quot;:&quot;Domain&quot;,&quot;bits&quot;:&quot;116.0&quot;,&quot;clan&quot;:&quot;CL0042&quot;,&quot;seq&quot;:{&quot;to&quot;:&quot;749&quot;,&quot;from&quot;:&quot;612&quot;,&quot;name&quot;:&quot;EMBOSS_001&quot;}}]
</pre></div>
</div>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>A local sequence tag and a [ID] block is the minimun requirment for EPTA to recognize a sequence.</p>
</div>
</div></blockquote>
</section>
<section id="taxonomy-information">
<h4>Taxonomy information<a class="headerlink" href="#taxonomy-information" title="Permalink to this headline">¶</a></h4>
<p>When using a ‘-tax’ flags to enable the taxonomy finding functionality, EPTA can fetch organism lineage
automatically from the NCBI database with a given protein ID, and the taxonomy information will be stored
in dataframe in the following format:</p>
<blockquote>
<div><div class="highlight-text notranslate"><div class="highlight"><pre><span></span>cellular organisms; Eukaryota; Sar; Stramenopiles; Bigyra; Opalozoa; Opalinata; Blastocystidae; Blastocystis; Blastocystis sp. subtypes; Blastocystis sp. subtype 1
</pre></div>
</div>
</div></blockquote>
<p>Once a sequence has organism lineage, it would be easy for show species or select proteins from organism under a certain classification in later process.</p>
<blockquote>
<div><p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax
</pre></div>
</div>
</div></blockquote>
</section>
<section id="protein-name-from-database">
<h4>Protein name from database<a class="headerlink" href="#protein-name-from-database" title="Permalink to this headline">¶</a></h4>
<p>EPTA can search protein name according to accession numbers from Entrez database.
To enable this function, add <strong>-name</strong> flag to the command line.</p>
<blockquote>
<div><p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name
</pre></div>
</div>
</div></blockquote>
</section>
<section id="run-pfamscan">
<h4>Run PfamScan<a class="headerlink" href="#run-pfamscan" title="Permalink to this headline">¶</a></h4>
<p>EPTA introduced PfamScan to allow users annotate protein domains on the tree.</p>
<p>Add <strong>‘-pfam’</strong> flag to the command line to run PfamScan and enable domain annotating.</p>
<p>Pfamscan search can provide name, accession number, hit sequence, envelope, e-value, bit score, and active site of a protein domain.
Which is a recommended way to obtain protein domain information.</p>
<p>Alongside <strong>‘-pfam’</strong> flag, <strong>‘-pev [E-value]’</strong> and <strong>‘-pas’</strong> [active sites] can be used to spicify a e-value and to enable active site functionality
in Pfamscan search. Meanwhile, both shortened domain data in the dataframe and a human-readable tsv file will be automatically generated after a
Pfamscan search in order to provide detailed information of protein domains.</p>
<p>We also introduced the ‘-em’ flag for users to provide an email address for receive protein messages from web services (for example, you are
temporarily banned).</p>
<blockquote>
<div><p>Example of Pfamscan result in dataframe:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>{&#39;Iron only hydrogenase large subunit, C-terminal domain; PF02906.14; evalue=1.3e-42; bits=146.1&#39;: &#39;1...186&#39;, &#39;Iron hydrogenase small subunit; PF02256.17; evalue=4.3e-26; bits=90.7&#39;: &#39;195...249&#39;, &#39;Flavodoxin; PF00258.25; evalue=2.5e-31; bits=108.8&#39;: &#39;276...415&#39;, &#39;FAD binding domain; PF00667.20; evalue=3.9e-42; bits=144.3&#39;: &#39;458...666&#39;, &#39;Oxidoreductase NAD-binding domain ; PF00175.21; evalue=2.2e-10; bits=41.1&#39;: &#39;701...800&#39;}
</pre></div>
</div>
<p>Example of Pfamscan details form:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>                 seq_name   seq_id     alignment_start  alignment_end  envelope_start  envelope_end  hmm_acc     hmm_name     hmm_desc                                               type   hmm_start  hmm_end  hmm_length  bit score  E-value  significance  clan  predicted_active_site_residues

i6j8l50wKOrP7Xpz GIQ79514.1 GIQ79514.1 1                186            1               186           PF02906.14  Fe_hyd_lg_C  Iron only hydrogenase large subunit, C-terminal domain Domain 104        243      243         146.1      1.3e-42  1             No_clan
i6j8l50wKOrP7Xpz GIQ79514.1 GIQ79514.1 195              249            194             249           PF02256.17  Fe_hyd_SSU   Iron hydrogenase small subunit                         Domain 2          56       56          90.7       4.3e-26  1             No_clan
i6j8l50wKOrP7Xpz GIQ79514.1 GIQ79514.1 276              415            276             415           PF00258.25  Flavodoxin_1 Flavodoxin                                             Domain 1          143      143         108.8      2.5e-31  1             CL0042
</pre></div>
</div>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name -pfam -pev <span class="m">10</span> -pas -em test@example.com
</pre></div>
</div>
</div></blockquote>
<div class="admonition caution">
<p class="admonition-title">Caution</p>
<p>According to the rule of the Pfamscan web service, one can submit at most 3000 sequences in a maximum of 30 batch jobs when using the Lite version.</p>
<p>Therefore, if you have more than 1500 sequences that need to run Pfamscan search, please do not run our tool again immediately
after keyboard interrupting a run that has already submit sequences to the server.</p>
</div>
</section>
</section>
</section>
<section id="multiple-sequence-alignment">
<h2>Multiple sequence alignment<a class="headerlink" href="#multiple-sequence-alignment" title="Permalink to this headline">¶</a></h2>
<p>EPTA introduced MAFFT and MUSCLE as multiple sequence alignment programs. The default
program is MAFFT, which could be changed in the configuration file.</p>
<p>After run multiple sequence alignment, trimAl was introduced to provide further sequence
trim.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>If both MAFFT and MUSCLE are set as default program, MAFFT is the prioritized one.</p>
</div>
<section id="mafft">
<h3>MAFFT<a class="headerlink" href="#mafft" title="Permalink to this headline">¶</a></h3>
<section id="run-mafft">
<h4>Run MAFFT<a class="headerlink" href="#run-mafft" title="Permalink to this headline">¶</a></h4>
<p>MAFFT will automatically run, do not need extra command line flags. However, users
still able to run MAFFT manually by adding <strong>‘-mafft’</strong> flag.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name -pfam -mafft
</pre></div>
</div>
<p>There are also several command line flags for changing paramters of MAFFT:</p>
<blockquote>
<div><p><strong>Matrix</strong>
Commnad line flag of matrix selecting is <strong>‘-matrix [Matrix Abbreviation]’</strong>,
identical to the command line flag ‘–bl’ or ‘–jtt’ of MAFFT.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name -pfam -mafft -matrix bl62
</pre></div>
</div>
<p><strong>Opening Score And Extension Score</strong>
Commnad line flag of opening score setting is <strong>‘-op [Number]’</strong>, identical to the command
line flag ‘–op’ of MAFFT.</p>
<p>Commnad line flag of extension score setting is <strong>‘-ep [Number]’</strong>, identical to the command
line flag ‘–ep’ of MAFFT.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name -pfam -mafft -op <span class="m">1</span>.53 -ep <span class="m">0</span>.123
</pre></div>
</div>
<p><strong>Tree Rebuilding Number</strong></p>
<p>Commnad line flag of tree rebuilding number is <strong>‘-retree [Number]’</strong>, identical to the command
line flag ‘–retree’ of MAFFT. This flag can determine the guide tree built times in the progressive
stage.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -em test@example.com -mafft -retree <span class="m">2</span>
</pre></div>
</div>
<p><strong>Max Iterate Number</strong></p>
<p>Commnad line flag of maximum iteration is <strong>‘-maxiterate [Number]’</strong>, identical to the command
line flag ‘–maxiterate’ of MAFFT. This flag can determine the cycles number of iterative refinement.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -em test@example.com -mafft -maxiterate <span class="m">2</span>
</pre></div>
</div>
<p><strong>Fast Fourier Transform Algorithm</strong></p>
<p>Commnad line flag of choose FFTS (Fast Fourier Transform) method is <strong>‘-ffts [Mode]’</strong>,
identical to the command line flag ‘–localpair’, ‘–genafpair’ and ‘–globalpair’ of MAFFT.
For each command, <strong>‘-ffts localpair’</strong> stands for the Smith-Waterman algorithm, <strong>‘-ffts genafpair’</strong>
stands for generalized affine gap cost, <strong>‘-ffts globalpair’</strong> stands for Needleman-Wunsch
algorithm.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -em test@example.com -mafft -ffts localpair
</pre></div>
</div>
</div></blockquote>
</section>
</section>
<section id="muscle">
<h3>MUSCLE<a class="headerlink" href="#muscle" title="Permalink to this headline">¶</a></h3>
<section id="run-muscle">
<h4>Run MUSCLE<a class="headerlink" href="#run-muscle" title="Permalink to this headline">¶</a></h4>
<p>To run MUSCLE as the multiple sequence alignment program, add <strong>-muscle</strong> flag to
the command line. MUSCLE will automatically run, do not need any parameter settings.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -muscle
</pre></div>
</div>
</section>
</section>
<section id="trimal">
<h3>trimAl<a class="headerlink" href="#trimal" title="Permalink to this headline">¶</a></h3>
<section id="run-trimal">
<h4>Run trimAl<a class="headerlink" href="#run-trimal" title="Permalink to this headline">¶</a></h4>
<p>As the default setting, trimAl automatically runs in the EPTA pipline. One can also
manually add <strong>-trimal</strong> flag to the command line to run trimAl.</p>
<p><strong>trimAl Run Mode</strong></p>
<p>Commnad line flag of selecting trimAl run mode <strong>‘-tmod [Mode]’</strong>. There are four
mode selectable, ‘automated1’, ‘gappyout’, ‘strict’, and ‘strictplus’, identical to
corresponding command of trimAl command line. Only ‘automated1’ mode is accessable in
the Lite mode.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -tmod automated1
</pre></div>
</div>
<p><strong>Remove Spurious Sequences</strong></p>
<p>The flag <strong>‘-rmss [Residue overlap/Sequence overlap]’</strong> able to remove spurious
sequences when triming multiple alignment sequences. <strong>Residue overlap</strong> is identical to
the flag ‘-resoverla’ of trimAl, in charge of keep “good positions” according to
a given minimum overlap of a positions with other positions in the column. <strong>Sequence overlap</strong>
is identical to the flag ‘-seqoverlap’ of trimAl, which means a minimum percentage of
“good positions” that a sequence must have in order to be conserved.</p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -rmss <span class="m">0</span>.75/80
</pre></div>
</div>
</section>
</section>
</section>
<section id="make-a-tree-from-alignment-file">
<h2>Make a tree from alignment file<a class="headerlink" href="#make-a-tree-from-alignment-file" title="Permalink to this headline">¶</a></h2>
<p>EPTA implements IQ-tree as the tree building program in the pipeline. When running
EPTA standalone version with large dataset (&gt;500), be aware that IQ-tree need corresponding
large memory space to run.</p>
<p>As the default setting, IQ-tree automatically runs in the EPTA pipline. One can also
manually add <strong>-iqtree</strong> flag to the command line to run IQ-tree.</p>
<p><strong>IQ-tree Run Mode</strong>
EPTA provide three run mode for users to select model: <strong>‘-iqmod TEST’</strong>, <strong>-iqmod TESTNEW</strong>
and <strong>-iqmod [A specific model]</strong>. They are based on IQ-tree’s command line, ‘TEST’ mode provides
automatic model selection in a basic model list, similarly, ‘TESTNEW’ automatically select model
from a advanced list. Users can also choose on known best model from the two list before. For more
details of models, please check the model list of IQ-tree:</p>
<p><a class="reference external" href="http://www.iqtree.org/doc/Tutorial#choosing-the-right-substitution-model">Model selection</a>
<a class="reference external" href="http://www.iqtree.org/doc/Substitution-Models">Substitution models</a></p>
<p>Command-line example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -iqmod TEST

epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -iqmod LG+I+G4
</pre></div>
</div>
<p><strong>Bootstrap</strong>
EPTA assessing branch support via ultrafast bootstrap of IQ-tree. The minimun number of
ultrafast bootstrap is 1000. The command flag of bootstrap is <strong>-boost [bootstrap numbr]</strong>, default
value set as 1000.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -iqmod TEST -boots <span class="m">1000</span>
</pre></div>
</div>
<p><strong>Reduce Runtime</strong>
To reduce runtime and computational burden, EPTA introduced <strong>-rcluster [percentage]</strong> flag, which implements
by IQ-tree, derived from the <em>relaxed hierarchical clustering algorithm</em>. This flag specify a percentage for the
relaxed clustering algorithm, in order to speed up the computation by reduce the percentage of partition schemes.
For example, <strong>-rcluster 10</strong> means only top 10% of partition schemes are considered in the running. This function
is turned off in the default setting.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -iqmod TEST -boots <span class="m">1000</span> -rcluster <span class="m">10</span>
</pre></div>
</div>
<p><strong>Tree Optimize</strong>
EPTA provides two command flags to enhance tree quality by increase run time. <strong>-mtree</strong> turns full tree search on to
increse the accuracy, and <strong>-bnni</strong> performs an additional step to further optimize UFBoot trees by nearest neighbor
interchange (NNI) based directly on bootstrap alignments. . One point needs to be awared is that both flags significantly
increase the run time.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -mafft -trimal -iqmod TEST -boots <span class="m">1000</span> -mtree -bnni
</pre></div>
</div>
</section>
<section id="annotate-a-tree-file">
<h2>Annotate a tree file<a class="headerlink" href="#annotate-a-tree-file" title="Permalink to this headline">¶</a></h2>
<p><strong>Annotate Taxonomy</strong>
By adding <strong>-tax</strong> flag in the command line, EPTA can automatically annotate taxonomy of all sequences’ species.
The maximum rank of the organism lineage EPTA can annotating on the phylogenetic tree is six. Any organism lineage
that longer than six rank will be shortened to that number. Given that the full organism lineage from NCBI commomly
starts with kingdom, EPTA intercepts the second to the last rank. If the full lineage longer than six rank (in the most cases)
, EPTA will intercept the second rank to the fifth rank and append the last rank then, which is the name of that specie.
For example:</p>
<blockquote>
<div><div class="highlight-text notranslate"><div class="highlight"><pre><span></span><span class="linenos">1</span>Full organism lineage:
<span class="linenos">2</span>cellular organisms; Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacterales; Enterobacteriaceae; Escherichia; Escherichia coli
<span class="linenos">3</span>
<span class="linenos">4</span>EPTA annotated organism lineage:
<span class="linenos">5</span>Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacterales; Enterobacteriaceae; Escherichia coli
</pre></div>
</div>
</div></blockquote>
<p>Correspondingly, if there are local sequence in the input file, pleas ensure that the full organism is attached.</p>
<p>Besides annotate taxonomic information on the tree image, EPTA is also capable to color tree branches depends on taxonomic
rank by adding <strong>-marktax</strong> in the command line.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -marktax
</pre></div>
</div>
<p><strong>Annotate Domain</strong>
By adding <strong>-pfam</strong> flag in the command line, EPTA can automatically annotate domains of all sequences. The top 10
frequent domain will be colored, and the rest domains will be annotated in gray. What’s more, for colored domains,
besides show domain names on each domain face, <strong>-leg</strong> command can make a lengend in the right bottom corner of the
image to illustrate domain names and corresponding color.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -pfam -leg
</pre></div>
</div>
<p><strong>Image Size</strong>
Command flag <strong>-xzoom [Magnification]</strong>, <strong>-yzoom [Magnification]</strong> are introduced to adjust the image size horizontally and
vertically. The basic magnification is 1. The default DPI of output png image is 300, and do not change with image size.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -xzoom <span class="m">1</span>.5 -yzoom <span class="m">1</span>.5
</pre></div>
</div>
<p><strong>Branch Setting</strong>
There are three attribute value of branches in EPTA, genetic distance, bootstrap value, and bifurcate number, corresponding
to the following commands: <strong>-bl</strong> (branch length), <strong>-bs</strong> (branch support), and <strong>-bif</strong> (bifurcation). None of them is enabled
in the defualt setting. Therefore, users need to enable corresponding command flag to show the value on tree image.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>epta -i ./fasta_files -o ./test -tax -name  -pfam -bl -bs -bif
</pre></div>
</div>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Manual</a><ul>
<li><a class="reference internal" href="#before-you-start">Before you start</a><ul>
<li><a class="reference internal" href="#configuration-file">Configuration file</a></li>
<li><a class="reference internal" href="#log-file">Log file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#make-a-tree-from-fasta-file">Make a tree from FASTA file</a><ul>
<li><a class="reference internal" href="#example-command-line">Example command-line</a></li>
<li><a class="reference internal" href="#input-file">Input file</a></li>
<li><a class="reference internal" href="#output-path">Output path</a></li>
<li><a class="reference internal" href="#check-point">Check point</a></li>
<li><a class="reference internal" href="#parse-fasta">Parse FASTA</a><ul>
<li><a class="reference internal" href="#sequence-recognition">Sequence recognition</a></li>
<li><a class="reference internal" href="#dataframe">Dataframe</a></li>
<li><a class="reference internal" href="#duplicate-headers">Duplicate headers?</a></li>
<li><a class="reference internal" href="#local-data">Local data?</a></li>
<li><a class="reference internal" href="#taxonomy-information">Taxonomy information</a></li>
<li><a class="reference internal" href="#protein-name-from-database">Protein name from database</a></li>
<li><a class="reference internal" href="#run-pfamscan">Run PfamScan</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#multiple-sequence-alignment">Multiple sequence alignment</a><ul>
<li><a class="reference internal" href="#mafft">MAFFT</a><ul>
<li><a class="reference internal" href="#run-mafft">Run MAFFT</a></li>
</ul>
</li>
<li><a class="reference internal" href="#muscle">MUSCLE</a><ul>
<li><a class="reference internal" href="#run-muscle">Run MUSCLE</a></li>
</ul>
</li>
<li><a class="reference internal" href="#trimal">trimAl</a><ul>
<li><a class="reference internal" href="#run-trimal">Run trimAl</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#make-a-tree-from-alignment-file">Make a tree from alignment file</a></li>
<li><a class="reference internal" href="#annotate-a-tree-file">Annotate a tree file</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="quickstart.html"
                        title="previous chapter">Quick Start</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="commandline.html"
                        title="next chapter">Commandline Reference</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/manual/manual.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="commandline.html" title="Commandline Reference"
             >next</a> |</li>
        <li class="right" >
          <a href="quickstart.html" title="Quick Start"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Easy Protein Tree Annotating  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Manual</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2021, Xuran Zhao.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.0.2.
    </div>
  </body>
</html>